### PR TITLE
fix(Traces Explorer): prevent duplicate API calls to query_range in traces explorer

### LINKED ISSUE #6308: FE[TRACES]: Query range api is triggered twice when trace explorer page is loaded.
Query range api is triggered twice when trace explorer page is loaded.

https://github.com/user-attachments/assets/76242bf3-08ee-4004-9a5b-46d4b7eba137

### PR DESCRIPTION
**Pull request by ahmadshaheer**
### Summary

- Memoize the query key to ensure that we don't have multiple query keys.
- Make API call only when the time range state is not loading (while the time range state is being updated).
- Added a check in the Filter component to avoid redundant redirection when the current query matches the prepared query.
- set `location.pathname` as the initial value for currentPathnameRef, absense of this causes `stagedQuery && location.pathname !== currentPathnameRef.current` to be truthy (since `currentPathnameRef.current` is null), and therefore causes duplicate triggering of `initQueryBuilderData(compositeQueryParam)`; in the useEffect of line ~784
    ```tsx
    if (!isValid) {
        redirectWithQueryBuilderData(validData);
    } else {
        initQueryBuilderData(compositeQueryParam);
    }
    ```

#### Related Issues / PR's
close #6308 

Previous PR: #6640

#### Screenshots

Before:

https://github.com/user-attachments/assets/7ec3b450-0fe3-43e9-843a-6d633f19d442

After:

https://github.com/user-attachments/assets/d88e5729-fa46-40ab-80cc-308e4cd07b1e

#### Affected Areas and Manually Tested Areas

----

> [!IMPORTANT]
> Prevent duplicate API calls in Traces Explorer by memoizing query keys and adding conditional checks in `index.tsx`, `Filter.tsx`, and `QueryBuilder.tsx`.
> 
>   - **Behavior**:
>     - Memoize `queryKey` in `index.tsx` to prevent duplicate API calls in Traces Explorer.
>     - API call in `useGetQueryRange` in `index.tsx` is now conditional on `!timeRangeUpdateLoading`.
>     - In `Filter.tsx`, added check to prevent redundant redirection if `currentQuery` matches `preparedQuery`.
>   - **State Management**:
>     - Initialize `currentPathnameRef` in `QueryBuilder.tsx` to prevent duplicate `initQueryBuilderData` calls.
>   - **Misc**:
>     - Minor refactoring in `index.tsx` and `QueryBuilder.tsx` for better readability and performance.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=SigNoz%2Fsignoz&utm_source=github&utm_medium=referral)<sup> for e76b7e24d900ff1214d0869ab4fa6effb2e92f9b. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: frontend/src/providers/QueryBuilder.tsx [hunk 1]
```diff
@@ -233,6 +243,7 @@ export function QueryBuilderProvider({
 				timeUpdated ? merge(currentQuery, newQueryState) : newQueryState,
 			);
 			setQueryType(type);
+			// debugger;
```
Comment (vikrantgupta25) [inline frontend/src/providers/QueryBuilder.tsx:246]: please remove this

File: frontend/src/providers/QueryBuilder.tsx [hunk 2]
```diff
@@ -792,6 +803,7 @@ export function QueryBuilderProvider({
 			initialQueriesMap.metrics,
 		);
 
+		// debugger;
```
Comment (vikrantgupta25) [inline frontend/src/providers/QueryBuilder.tsx:806]: remove this as well

File: frontend/src/providers/QueryBuilder.tsx [hunk 3]
```diff
@@ -95,7 +95,17 @@ export function QueryBuilderProvider({
 	const urlQuery = useUrlQuery();
 	const history = useHistory();
 	const location = useLocation();
-	const currentPathnameRef = useRef<string | null>(null);
+	/**
+	 // set location.pathname as the initial value for currentPathnameRef, absense of this causes `stagedQuery && location.pathname !== currentPathnameRef.current` to be truthy (since currentPathnameRef.current is null), and therefore causes duplicate triggering of initQueryBuilderData(compositeQueryParam); in the useEffect of line ~784
```
Comment (vikrantgupta25) [inline frontend/src/providers/QueryBuilder.tsx:99]: let's add this as PR description. Inline comment will pollute the provider here

**Review by vikrantgupta25 — COMMENTED** [review]
Remove the debuggers, rest looks good !

**Review by YounixM — DISMISSED** [review]

File: frontend/src/providers/QueryBuilder.tsx (hunk 1 above)
Comment (ahmadshaheer) [reply frontend/src/providers/QueryBuilder.tsx:246]: Done

File: frontend/src/providers/QueryBuilder.tsx (hunk 2 above)
Comment (ahmadshaheer) [reply frontend/src/providers/QueryBuilder.tsx:806]: Done

File: frontend/src/providers/QueryBuilder.tsx (hunk 3 above)
Comment (ahmadshaheer) [reply frontend/src/providers/QueryBuilder.tsx:99]: Sure, :+1:

**Review by SagarRajput-7 — DISMISSED** [review]

**Review by vikrantgupta25 — DISMISSED** [review]

**Review by vikrantgupta25 — APPROVED** [review]

**Pull request merged**